File: backend/Models/Network.cs
namespace Graphic_Neural_Network.backend.Models
{
    public class Network
    {
        public List<Layer> Layers { get; set; }
        public int MaxHiddenLayers { get; set; } = 4;
        public int MaxNeurons { get; set; } = 10;
        public int InputsNumber { get; set; } = 2;
        public int OutputsNumber { get; set; } = 2;
        public int HiddenLayerCount { get; set; } = 4;
        public Dictionary<string, bool> Collapsed { get; set; } = new Dictionary<string, bool> {{ "Network", false },{ "Sliders", false} };

        public Network()
        {
            Layers = new List<Layer>
            {
                new Layer(4, 2),
                new Layer(2, 4),
                new Layer(10, 2),
                new Layer(7, 10),
                new Layer(2, 10)
            };
        }

        public void AddNode(int layerNumber)
        {
            if (Layers[layerNumber].NeuronsNumber >= MaxNeurons || layerNumber >= HiddenLayerCount)
                return;

            Layers[layerNumber].AddNeuron();
            if (layerNumber < HiddenLayerCount)
            {
                foreach (var weightRow in Layers[layerNumber + 1].Weights)
                {
                    weightRow.Add(0);
                }
            }
        }

        public void RemoveNode(int layerNumber, int neuronNumber)
        {
            if (layerNumber >= HiddenLayerCount)
                return;

            if (Layers[layerNumber].NeuronsNumber == 1)
            {
                RemoveLayer(layerNumber);
                return;
            }

            Layers[layerNumber].RemoveNeuron(neuronNumber);
            foreach (var weightRow in Layers[layerNumber + 1].Weights)
            {
                weightRow.RemoveAt(neuronNumber);
            }
        }

        public void AddLayer(int index)
        {
            if (HiddenLayerCount == MaxHiddenLayers)
                return;

            var newLayer = index == 0
                ? new Layer(1, 2)
                : new Layer(1, Layers[index].NeuronsNumber);

            Layers.Insert(index, newLayer);
            Layers[index + 1].PrevLayerNeuronsNumber = 1;
            Layers[index + 1].Weights = new List<List<double>>();
            for (int i = 0; i < Layers[index + 1].NeuronsNumber; i++)
            {
                Layers[index + 1].Weights.Add(new List<double> { 0 });
            }

            HiddenLayerCount++;
        }

        public void RemoveLayer(int layerNumber)
        {
            if (HiddenLayerCount == 1 || HiddenLayerCount == layerNumber)
                return;

            Layers[layerNumber + 1].PrevLayerNeuronsNumber = Layers[layerNumber].PrevLayerNeuronsNumber;
            Layers[layerNumber + 1].Weights = new List<List<double>>();
            for (int i = 0; i < Layers[layerNumber + 1].NeuronsNumber; i++)
            {
                Layers[layerNumber + 1].Weights.Add(new List<double>(new double[Layers[layerNumber].PrevLayerNeuronsNumber]));
            }

            Layers.RemoveAt(layerNumber);
            HiddenLayerCount--;
        }

        public void UpdateBiases(int layerNumber, int neuronNumber, double newBias)
        {
            Layers[layerNumber].UpdateBias(neuronNumber, newBias);
        }

        public void UpdateWeights(int layerNumber, int neuronNumber, int inputNumber, double newWeight)
        {
            Layers[layerNumber].UpdateWeight(neuronNumber, inputNumber, newWeight);
        }

        public void UpdateClassCount(int count)
        {
            if ((count == -1 && OutputsNumber == 2) || (count == 1 && OutputsNumber == 5))
                return;

            OutputsNumber += count;

            if (count < 0)
            {
                Layers[HiddenLayerCount].RemoveNeuron(Layers[HiddenLayerCount].NeuronsNumber - 1);
            }
            else
            {
                Layers[HiddenLayerCount].AddNeuron();
            }
        }

        public void UpdateCollapsed(string key , bool value) 
        {
            if (Collapsed.ContainsKey(key))
            {
                Collapsed[key] = value;
            }
        }
    }
}

File: frontend/src/models/network/Network.ts
import Layer from "./Layer";


export default class Network {
    layers: Layer[];
    //
    maxHidenLayers: number = 4;
    maxNeurons: number = 10;
    //
    inputsNumber: number = 2;
    outputsNumber: number = 2;
    //
    hidenLayerCount: number = 1;

    constructor() {
        this.layers = [
            new Layer(3, 2),

            new Layer(2, 3)]
    }

    regenerateNetwork() {
        this.layers.forEach((layer: Layer) => {
            layer.biases = Array.from({ length: layer.neuronsNumber }, () => Math.random() * 2 - 1);
            layer.weights = Array.from({ length: layer.neuronsNumber }, () =>
                Array.from({ length: layer.prevLayerNeuronsNumber }, () => Math.random() * 2 - 1));

        });
    }


    // add node to layer
    addNode(layerNumber: number) {
        if (this.layers[layerNumber].neuronsNumber >= this.maxNeurons) return;
        if (layerNumber >= this.hidenLayerCount) return;
        this.layers[layerNumber].addNeuron();
        if (layerNumber < this.hidenLayerCount)
            for (let i = 0; i < this.layers[layerNumber + 1].weights.length;i++)
            this.layers[layerNumber + 1].weights[i].push(0);

    }

    // remove node from layer
    removeNode(layerNumber: number, neuronsNumber: number) {
        if (layerNumber >= this.hidenLayerCount) return;
        if (this.layers[layerNumber].neuronsNumber == 1) {
            this.removeLayer(layerNumber);
            return;
        }
        this.layers[layerNumber].removeNeuron(neuronsNumber);
        //
        for (let i = 0; i < this.layers[layerNumber + 1].weights.length; i++)
            this.layers[layerNumber + 1].weights[i].splice(neuronsNumber, 1);


    }

    // add new layer
    addLayer(index: number) {
        if (this.hidenLayerCount == this.maxHidenLayers) return;
        if (index == 0)
            this.layers.splice(0, 0, new Layer(1, 2));
        else
            this.layers.splice(index, 0, new Layer(1, this.layers[index].neuronsNumber));
        // update next layer
        this.layers[index + 1].prevLayerNeuronsNumber = 1;
        this.layers[index + 1].weights = Array.from(
            { length: this.layers[index + 1].neuronsNumber },
            () => new Array(1).fill(0)
        );

        this.hidenLayerCount++;
    }

    // remove layer | should be ony called when layer has one node
    removeLayer(laterNumber: number) {
        if (this.hidenLayerCount == 1) return;
        // cant remove last
        if (this.hidenLayerCount == laterNumber) return;
        // update next layer baies
        this.layers[laterNumber + 1].prevLayerNeuronsNumber = this.layers[laterNumber].prevLayerNeuronsNumber;
        // update next layer weights
        this.layers[laterNumber + 1].weights = new Array(this.layers[laterNumber + 1].neuronsNumber).fill(new Array(this.layers[laterNumber].prevLayerNeuronsNumber).fill(0));
        // remove layer
        this.layers.splice(laterNumber, 1);
        this.hidenLayerCount--;
    }

    updateBaies(layerNumber: number, neuronNumber: number, newBias: number) {
        this.layers[layerNumber].updateBias(neuronNumber, newBias);
    }

    updateWeight(layerNumber: number, neuronNumber: number, inputNumber: number, newWeight: number) {
        this.layers[layerNumber].upadateWeight(neuronNumber, inputNumber, newWeight);
    }

    updateOutputsNumber(count: number) {
        if (count == -1 && this.outputsNumber == 2) return;
        if (count == 1 && this.outputsNumber == 5) return;
        this.outputsNumber += count;
        if(count < 0 )
            this.layers[this.hidenLayerCount].removeNeuron(this.layers[this.hidenLayerCount].neuronsNumber-1);
        else
            this.layers[this.hidenLayerCount].addNeuron();
    }

  
    activationFunctiond(inputs: number[]) { return inputs.map(x => 1 / (1 + Math.exp(-x)))}


    predict(inputs: number[], activationFunction: (inputs: number[]) => number[]) {
        let outputs = inputs;
        this.layers.forEach((layer: Layer) => {
            outputs = layer.layerPredict(outputs)!;
        });
        return this.activationFunctiond(outputs);
    }


   
}


  